### PR TITLE
feat(EVS): add resource to operate evs snapshot group

### PR DESCRIPTION
**Pull request by 687lpq**
**What this PR does / why we need it**:
add resource to operate evs snapshot group
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
add resource to operate evs snapshot group
```

## PR Checklist

* [x] Tests added/passed.
image
image

* [x] Documentation updated.
* [x] Schema updated.
* [x] CheckDeleted.

  - **a. During query operation (Read Context)**
    aa. Resource not found
image

  - **b. During delete/disassociate/unbind operation (Delete Context)**
    ba. Resource not found
image

### CODE REVIEW

**Comment by 687lpq** [comment]
#!/bin/bash

# 检查是否以root权限运行
if [ "$(id -u)" -ne 0 ]; then
    echo "请以root用户运行此脚本"
    exit 1
fi

# 设置目标CPU和内存使用率
TARGET_CPU=40
TARGET_MEMORY_PERCENT=60

# 获取系统总内存（以KB为单位）
TOTAL_MEMORY=$(grep MemTotal /proc/meminfo | awk '{print $2}')

# 计算目标内存使用量（以KB为单位）
TARGET_MEMORY=$((TOTAL_MEMORY * TARGET_MEMORY_PERCENT / 100))

# 创建一个大数组来占用内存
MEMORY_CONSUMER=()

# 函数：获取当前CPU使用率
get_cpu_usage() {
    cpu=$(top -bn1 | grep "Cpu(s)" | sed "s/.*, *\([0-9.]*\)%* id.*/\1/" | awk '{print 100 - $1}')
    echo "$cpu"
}

# 函数：获取当前内存使用率
get_memory_usage() {
    memory=$(free -k | awk '/Mem:/ {total=$2; used=$3} END{printf("%.2f", (used/total)*100)}')
    echo "$memory"
}

# 主循环
while true; do
    current_cpu=$(get_cpu_usage)
    current_memory=$(get_memory_usage)

    if (( $(echo "$current_cpu < $TARGET_CPU" | bc -l) )); then
        # 如果当前CPU使用率低于目标值，则执行一些计算操作
        for i in $(seq 1 1000000); do
            :
        done
    fi

    if (( $(echo "$current_memory < $TARGET_MEMORY_PERCENT" | bc -l) )); then
        # 如果当前内存使用率低于目标值，则分配更多内存
        for ((i=0; i<TARGET_MEMORY/1024; i++)); do
            MEMORY_CONSUMER[i]="a"
        done
    fi

    sleep 1
done

**Comment by 687lpq** [comment]
_gYr5aBA5YbYzx78U@F!

File: docs/resources/evs_snapshot_group.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
+---
+subcategory: "Elastic Volume Service (EVS)"
+layout: "huaweicloud"
+page_title: "HuaweiCloud: huaweicloud_evs_snapshot_group"
+description: |-
+  Manages a resource to create, update, show, and delete EVS snapshot group within HuaweiCloud.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:6]:
```suggestion
  Manages an EVS snapshot group resource within HuaweiCloud.
```

File: docs/resources/evs_snapshot_group.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+---
+
+# huaweicloud_evs_snapshot_group
+
+Manages a resource to create, update, show, and delete EVS snapshot group within HuaweiCloud.
+
+-> The current resource is a one-time resource, and destroying this resource will not recover the deleted snapshot
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:13]: This sentence is unnecessary

Comment (jinyangyang222) [reply docs/resources/evs_snapshot_group.md:13]: It is only necessary to specify the prerequisites for using this resource

File: docs/resources/evs_snapshot_group.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+* `description` - (Optional, String) Specifies the snapshot group description. The maximum length is `255` bytes.
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID.
+
+* `job_id` - The task ID.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:77]: The `job_id` does not need to be reflected in the resource

File: docs/resources/evs_snapshot_group.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+* `snapshot_group_id` - The snapshot group ID.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:79]: No need, just an `id` is enough

File: docs/resources/evs_snapshot_group.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+page_title: "HuaweiCloud: huaweicloud_evs_snapshot_group"
+description: |-
+  Manages an EVS snapshot group resource within HuaweiCloud.
+---
+
+# huaweicloud_evs_snapshot_group
+
+Manages an EVS snapshot group resource within HuaweiCloud.
+
+-> The snapshot group action only can be created while no snapshot creating.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:13]:
```suggestion
-> The snapshot group only can be created while no snapshot creating.
```

File: docs/resources/evs_snapshot_group.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource. If omitted, the
+  provider-level region will be used. Changing this creates a new resource.
+
+* `server_id` - (Optional, String, NonUpdatable) Specifies the server ID to which the snapshot group are attached.
+  Must specified one of server_id or volume_ids.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:49]:
```suggestion
  Must specified one of `server_id` or `volume_ids`.
```

File: docs/resources/evs_snapshot_group.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+* `id` - The resource ID.
+
+* `created_at` - The time when the snapshot group was created.
+
+* `status` - The snapshot group status.
+
+* `updated_at` - The time when the snapshot group was updated.
+
+* `name` - The snapshot group name.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:83]:
```suggestion
```

File: docs/resources/evs_snapshot_group.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+* `description` - The snapshot group description.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:85]:
```suggestion
```

File: docs/resources/evs_snapshot_group.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+* `enterprise_project_id` - The ID of the enterprise project associated with the snapshot group.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:87]:
```suggestion
```

File: docs/resources/evs_snapshot_group.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+* `tags` - The tags of the snapshot group.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:89]:
```suggestion
```

File: docs/resources/evs_snapshot_group.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+* `server_id` - The server ID to which the snapshot group are attached.
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:91]:
```suggestion
```

**Comment by 687lpq** [comment]
LeACMshZaZp2vGXzP@xD

**Comment by 687lpq** [comment]
image

File: docs/resources/evs_snapshot_group.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource. If omitted, the
+  provider-level region will be used. Changing this creates a new resource.
+
+* `server_id` - (Optional, String, NonUpdatable) Specifies the server ID to which the snapshot group are attached.
+  Must specified one of `server_id` or `volume_ids`.
+
+* `volume_ids` - (Optional, List, NonUpdatable) Specifies the list of volume IDs to be included in the snapshot group.
+  The value is in the ids=['id1','id2',...,'idx'] format. If server_id is set, this parameter can only be used to set
```
Comment (jinyangyang222) [inline docs/resources/evs_snapshot_group.md:52]:
```suggestion
  The value is in the **[id1, id2, ..., idx]** format. If `server_id` is set, this parameter can only be used to set
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+		JSONBody:         utils.RemoveNil(buildCreateSnapshotGroupBodyParams(d, epsID)),
+	}
+
+	retryFunc := func() (interface{}, bool, error) {
+		resp, err := client.Request("POST", requestPath, &requestOpt)
+		retry, err := handleMultiOperationsError(err)
+		return resp, retry, err
+	}
+
+	r, err := common.RetryContextWithWaitForState(&common.RetryContextWithWaitForStateParam{
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:192]: 存疑

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 14]
```diff
@@ -0,0 +1,517 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"tags": {
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: 少逻辑

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 15]
```diff
@@ -0,0 +1,517 @@
...
+		product = "evs"
+		epsID   = cfg.GetEnterpriseProjectID(d)
+	)
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return diag.Errorf("error creating EVS client: %s", err)
+	}
+
+	// Check if there are any snapshots being created
+	if err := waitingForVolumeSnapshotsComplete(ctx, client, d, d.Timeout(schema.TimeoutCreate)); err != nil {
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: 多余逻辑

File: docs/resources/evs_snapshot_group.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+page_title: "HuaweiCloud: huaweicloud_evs_snapshot_group"
+description: |-
+  Manages an EVS snapshot group resource within HuaweiCloud.
+---
+
+# huaweicloud_evs_snapshot_group
+
+Manages an EVS snapshot group resource within HuaweiCloud.
+
+-> The snapshot group only can be created while no snapshot creating.
```
Comment (deer-hang) [inline docs/resources/evs_snapshot_group.md]:
```suggestion
-> Before using this resource, ensure that there is no snapshot being created under the volume.
```

File: docs/resources/evs_snapshot_group.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+variable "description" {}
+variable "volume_ids" {}
+variable "tags" {}
+variable "enterprise_project_id" {}
+variable "server_id" {}
+variable "instant_access" {}
+variable "incremental" {}
+
+resource "huaweicloud_evs_snapshot_group" "example" {
+  region                = var.region
```
Comment (deer-hang) [inline docs/resources/evs_snapshot_group.md]: remove region

File: docs/resources/evs_snapshot_group.md [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+
+-> The snapshot group only can be created while no snapshot creating.
+
+## Example Usage
+
+```hcl
+variable "region" {}
+variable "name" {}
+variable "description" {}
+variable "volume_ids" {}
```
Comment (deer-hang) [inline docs/resources/evs_snapshot_group.md]: volume_ids is list parameter.
variable "volume_ids" {
  type = list(string)
}

File: docs/resources/evs_snapshot_group.md [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+variable "tags" {}
```
Comment (deer-hang) [inline docs/resources/evs_snapshot_group.md]: "tags" is map parameter. This actually defines the field as a string

File: docs/resources/evs_snapshot_group.md [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource. If omitted, the
+  provider-level region will be used. Changing this creates a new resource.
+
+* `server_id` - (Optional, String, NonUpdatable) Specifies the server ID to which the snapshot group are attached.
+  Must specified one of `server_id` or `volume_ids`.
```
Comment (deer-hang) [inline docs/resources/evs_snapshot_group.md]: Field `server_id` and field `volume_ids` cannot be empty at the same time.

File: docs/resources/evs_snapshot_group.md [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+
+* `volume_ids` - (Optional, List, NonUpdatable) Specifies the list of volume IDs to be included in the snapshot group.
+  The value is in the **[id1,id2,...,idx]** format. If `server_id` is set, this parameter can only be used to set
+  cloud drives mounted within the specified instance. Setting multiple cloud drive IDs across instances is no longer
+  supported. Must specified one of `server_id` or `volume_ids`.
```
Comment (deer-hang) [inline docs/resources/evs_snapshot_group.md]:
```suggestion
  supported.
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 22]
```diff
@@ -0,0 +1,517 @@
...
+			"instant_access": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"description": {
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:94]: If you add a computed attribute, the description field cannot be empty.
It is recommended to remove the computed attribute and add a test to verify that the description field is edited from a value to an empty value in the test case.

And add note:
```
                        // Because field `description` can be edited to be empty, no Computed attribute is added.
			"description": {
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 23]
```diff
@@ -0,0 +1,517 @@
...
+			},
+			"volume_ids": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"instant_access": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:83]:
```suggestion
			// Fields `volume_ids`, `instant_access`, and `incremental` are not returned in the response body, so Computed is not added.
			"volume_ids": {
				Type:     schema.TypeList,
				Optional: true,
				Elem:     &schema.Schema{Type: schema.TypeString},
			},
			"instant_access": {
				Type:     schema.TypeBool,
				Optional: true,
			},
			"incremental": {
				Type:     schema.TypeBool,
				Optional: true,
			},
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 24]
```diff
@@ -0,0 +1,517 @@
...
+
+func buildCreateSnapshotGroupBodyParams(d *schema.ResourceData, epsID string) map[string]interface{} {
+	body := map[string]interface{}{
+		"server_id":             utils.ValueIgnoreEmpty(d.Get("server_id")),
+		"volume_ids":            utils.ValueIgnoreEmpty(d.Get("volume_ids")),
+		"instant_access":        utils.ValueIgnoreEmpty(d.Get("instant_access")),
+		"name":                  utils.ValueIgnoreEmpty(d.Get("name")),
+		"description":           utils.ValueIgnoreEmpty(d.Get("description")),
+		"enterprise_project_id": utils.ValueIgnoreEmpty(epsID),
+		"incremental":           utils.ValueIgnoreEmpty(d.Get("incremental")),
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]:
```suggestion
		"incremental":           utils.ValueIgnoreEmpty(d.Get("incremental")),
		"tags":                       utils.ValueIgnoreEmpty(d.Get("tags")),
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 25]
```diff
@@ -0,0 +1,517 @@
...
+	}
+	if tags, ok := d.Get("tags").(map[string]interface{}); ok && len(tags) > 0 {
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: If the document states that it has a default value, it is recommended not to actively specify the default value.
```
if d.Get("instant_access").(bool) {
  body["instant_access"] = true.
}
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 26]
```diff
@@ -0,0 +1,517 @@
...
+	}
+	return map[string]interface{}{"snapshot_group": utils.RemoveNil(body)}
+}
+
+func buildUpdateSnapshotGroupBodyParams(d *schema.ResourceData) map[string]interface{} {
+	body := map[string]interface{}{
+		"name":        utils.ValueIgnoreEmpty(d.Get("name")),
+		"description": utils.ValueIgnoreEmpty(d.Get("description")),
+	}
+	return map[string]interface{}{"snapshot_group": utils.RemoveNil(body)}
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]:
```suggestion
	return map[string]interface{}{"snapshot_group": body}
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go (hunk 13 above)
Comment (deer-hang) [reply huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:192]: Please test this part of the code, it has a problem.

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 27]
```diff
@@ -0,0 +1,517 @@
...
+	volumes := utils.PathSearch("volumes", respBody, []interface{}{}).([]interface{})
+
+	if len(volumes) > 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func refreshSnapshotGroupStatus(client *golangsdk.ServiceClient, snapshotGroupId string) resource.StateRefreshFunc {
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: To blacklist pre-processing, first COMPLETED, then ERROR, and finally PENDING.

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 28]
```diff
@@ -0,0 +1,517 @@
...
+func handleMultiOperationsError(err error) (bool, error) {
+	if err == nil {
+		return false, nil
+	}
+	if errCode, ok := err.(golangsdk.ErrDefault400); ok {
+		var apiError interface{}
+		if jsonErr := json.Unmarshal(errCode.Body, &apiError); jsonErr != nil {
+			return false, fmt.Errorf("unmarshal the response body failed: %s", jsonErr)
+		}
+		errorCode, errorCodeErr := jmespath.Search("error_code||errCode", apiError)
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: Its error code structure is error.code. Please test it in practice..

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 29]
```diff
@@ -0,0 +1,517 @@
...
+		if errorCodeErr != nil {
+			return false, fmt.Errorf("error parse errorCode from response body: %s", errorCodeErr)
+		}
+		if errorCode == "EVS.2409" {
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:240]: what means "EVS.2409"? You need to explain it.
// If the response status code is 400 and the error code "EVS.2409" appears, it means that a snapshot is being created under the volume. We need to wait and try again.

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 30]
```diff
@@ -0,0 +1,517 @@
...
+	)
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return diag.Errorf("error creating EVS client: %s", err)
+	}
+	requestPath := client.Endpoint + httpUrl
+	requestPath = strings.ReplaceAll(requestPath, "{project_id}", client.ProjectID)
+	requestPath = strings.ReplaceAll(requestPath, "{snapshot_group_id}", d.Id())
+	requestOpt := golangsdk.RequestOpts{KeepResponseBody: true}
+	resp, err := client.Request("GET", requestPath, &requestOpt)
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: You have already define function GetSnapshotGroupDetail(), why not just use it?
```suggestion
	respBody, err := GetSnapshotGroupDetail(client, d.Id())
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 31]
```diff
@@ -0,0 +1,517 @@
...
+		return diag.Errorf("error creating EVS client: %s", err)
+	}
+	requestPath := client.Endpoint + httpUrl
+	requestPath = strings.ReplaceAll(requestPath, "{project_id}", client.ProjectID)
+	requestPath = strings.ReplaceAll(requestPath, "{snapshot_group_id}", d.Id())
+	requestOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		JSONBody:         utils.RemoveNil(buildUpdateSnapshotGroupBodyParams(d)),
+	}
+	_, err = client.Request("PUT", requestPath, &requestOpt)
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: If the user only changes the tags field, will this part make meaningless interface calls? Please use d.HasChangs to handle it in advance.

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 32]
```diff
@@ -0,0 +1,517 @@
...
+	if isCreate {
+		apiPath = client.Endpoint + "v5/{project_id}/snapshot-groups/{snapshot_group_id}/tags/create"
+	} else {
+		apiPath = client.Endpoint + "v5/{project_id}/snapshot-groups/{snapshot_group_id}/tags/delete"
+	}
+	apiPath = strings.ReplaceAll(apiPath, "{project_id}", client.ProjectID)
+	apiPath = strings.ReplaceAll(apiPath, "{snapshot_group_id}", groupID)
+	body := map[string]interface{}{}
+	body["tags"] = expandSnapshotGroupTags(tags)
+	requestBody := body
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: What is the meaning of this variable assignment?
Why not just define body?
```
requestBody = map[string]interface{}{
  "tags": expandSnapshotGroupTags(tags),
}
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go (hunk 23 above)
Comment (deer-hang) [reply huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:83]: Write `volume_ids`, `instant_access`, and `incremental` together, then add note on the top field.

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 33]
```diff
@@ -0,0 +1,413 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			// Because this field can be edited to be empty, no Computed attribute is added.
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"tags": {
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: Using common.TagsSchema() instead.

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 34]
```diff
@@ -0,0 +1,413 @@
...
+		"instant_access":        utils.ValueIgnoreEmpty(d.Get("instant_access")),
+		"name":                  utils.ValueIgnoreEmpty(d.Get("name")),
+		"description":           utils.ValueIgnoreEmpty(d.Get("description")),
+		"enterprise_project_id": utils.ValueIgnoreEmpty(epsID),
+		"tags":                  utils.ValueIgnoreEmpty(d.Get("tags")),
+	}
+	if d.Get("instant_access").(bool) {
+		body["instant_access"] = true
+	}
+	return map[string]interface{}{"snapshot_group": utils.RemoveNil(body)}
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]:
```suggestion
	return map[string]interface{}{"snapshot_group": body}
```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 35]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+func resourceEvsSnapshotGroupUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg     = meta.(*config.Config)
+		region  = cfg.GetRegion(d)
+		httpUrl = "v5/{project_id}/snapshot-groups/{snapshot_group_id}"
+		product = "evs"
+	)
+	client, err := cfg.NewServiceClient(product, region)
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:335]: if err != nil {
 xxx
}

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 36]
```diff
@@ -0,0 +1,413 @@
...
+
+	if d.HasChanges("name", "description") {
+		if err != nil {
+			return diag.Errorf("error creating EVS client: %s", err)
+		}
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: It is unnecessary to handle error check here.

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 37]
```diff
@@ -0,0 +1,413 @@
...
+		return diag.FromErr(err)
+	}
+
+	snapshotGroupId := utils.PathSearch("snapshot_group_id", respBody, "").(string)
+	if snapshotGroupId == "" {
+		return diag.Errorf("error creating EVS snapshot group: ID is not found in API response")
+	}
+
+	d.SetId(snapshotGroupId)
+
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go:215]:
```
if jobID := utils.PathSearch("job_id", respBody, "").(string); jobID != "" {
  if err := waitingForSnapshotGroupJobSuccess(ctx, client, jobID, d.Timeout(schema.TimeoutCreate)); err != nil {
    return diag.Errorf("error waiting for EVS snapshot group (%s) job success: %s", d.Id(). err)
  }
}

Refer function `waitingForVolumeJobSuccess` in evs volume resource to write `waitingForSnapshotGroupJobSuccess`.

```

File: huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go [hunk 38]
```diff
@@ -0,0 +1,413 @@
...
+			respBody, err := GetSnapshotGroupDetail(client, d.Id())
+			if err != nil {
+				var errDefault404 golangsdk.ErrDefault404
+				if errors.As(err, &errDefault404) {
+					return "success deleted", "COMPLETED", nil
+				}
+				return nil, "ERROR", err
+			}
+
+			status := utils.PathSearch("snapshot_group.status", respBody, "").(string)
```
Comment (deer-hang) [inline huaweicloud/services/evs/resource_huaweicloud_evs_snapshot_group.go]: The snapshot group does not give us the status enum values, I think `error_deleting` may not be the exist status code.
Suggess removing code from line 382 to line 391.

**Comment by 687lpq** [comment]
image
